コメント機能課題の提出

YukiWatanabe824 opened the pull request:
提出いたします。
ご確認お願いいたします。

YukiWatanabe824 committed:
scaffoldを実行

YukiWatanabe824 committed:
日報機能の追加

YukiWatanabe824 committed:
コメント機能の実装

YukiWatanabe824 committed:
i18nの対応、不要なファイルやコードの削除

YukiWatanabe824 committed:
不要なファイルの削除

YukiWatanabe824 committed:
rubocopの修正

YukiWatanabe824 committed:
erblintの修正

betachelsea commented on app/controllers/books_controller.rb:
```diff
@@ -11,7 +11,11 @@ def index
+  def show
+    set_commentable
+    @comments = Book.find_by(id: params[:id]).comments.all
+    @comment = @commentable.comments.build
```
[nits] 細かい点ですが、いくつか 🙏 
- `Book.find_by(id: params[:id])` は、`@book` で代用できそうです。
- `@commentable` も `@book` と等しいなと思いました。
- この辺りは好みの問題ですが、view側で直接 `@book.comments.all` などと呼んでしまうのも一手です。

betachelsea commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,66 @@
+
+  # GET /comments/1/edit
+  def edit
+    redirect_to comments_path if @comment.user.id != current_user.id
```
[nits] コメントの編集の実装は無いとのことで、ルーティングとアクションも無い機能は未定義にしておくと、将来困らなくて便利です！ (存在しない機能も、コードが書いてあると、N年後に見た時に保守対象なのか or 消しても良いのかという判断が必要となり、意外と苦戦します 😅  )

betachelsea commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,66 @@
+  private
+
+  def set_book_or_report
+    @commentable = if request.url.include?('reports')
```
[nits] pathがBookとReport用の2種類ある状況で、同一の CommentsController が使われているため、こうする必要があるようですね。別案として、Controller自体を分けるという手段もあるということを共有しておきます。 🙏

YukiWatanabe824 replied:
指摘ありがとうございます。

`Controller自体を分ける` という部分について
`concernで共通化したルーティングをBookとReportで使用する形`から、`Commentのルーティング(コントローラ)をBookとReportの名前空間でグループ化`する形に変更しました。
これでBookとReportで別個の処理は各モデル配下のControllerに処理を記載し、共通処理はそのままCommentControllerを活用しています。

ここで質問です。
モデルごとにControllerを分ける事自体は途中で初回提出前にも思いついたのですが、プラクティスの終了条件に`ポリモーフィック関連を使ってbookに付くコメントとreportに付くコメントを共通のコードで動くようにする。`とあり共通化させた経緯があります。
ポリモーフィック関連で処理を共通化する/しないの線引きはなにかあるでしょうか？

betachelsea replied:
> ポリモーフィック関連で処理を共通化する/しないの線引き

今回の終了条件の意図としては、コメントの仕組みを考慮する際に、`BookComment` `ReportComment` のようにテーブル・モデルを別物として作成しないようにしてください、という意味合いだと考えます。
処理の共通化する/しないの線引きは、それぞれのプロジェクト、TPOによるというところなので確実に守れば良いルールは無いですが、"ポリモーフィック関連" というのがテーブルの構成パターンに関係する用語ですので、そのあたりに注力して考えていただくのが良いかもしれません。また、共通化しようとしたときに無理がある/無理する必要に迫られるような部分が出てきた時は、無理せずにわける、というのが良いかなと思います。前回の指摘時の "無理しているな" と思った点としては、 `request.url` から report or book を判別しようとしていたところでした。

また、実際のプロジェクトだったら、迷った時は実装前にチームメンバーに確認をとるのも一手ですね。 😄

betachelsea commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,66 @@
+    @comment = Comment.new(comment_params)
+    respond_to do |format|
+      if @comment.save
+        format.html { redirect_to [@commentable], notice: t('controllers.common.notice_create', name: Comment.model_name.human) }
```
[nits] レスポンスがhtmlのみを想定している場合は、`respond_to do 〜 end` や、 `format.html` の表記は無くても大丈夫です。

```ruby
@comment = Comment.new(comment_params)
if @comment.save
  redirect_to [@commentable], notice: t('controllers.common.notice_create', name: Comment.model_name.human)
else
#...略...
```

また、コメントでもjsonの存在が示唆されているので（ `# POST /comments or /comments.json` ）、修正しておくと丁寧ですね 👍

betachelsea commented on app/controllers/reports_controller.rb:
```diff
@@ -0,0 +1,76 @@
+
+  # DELETE /reports/1 or /reports/1.json
+  def destroy
+    @report.destroy
```
[MUST]  `日報を更新・削除できるのはその日報を投稿した本人のみ` の必須要件の対応に失敗している様子です。確認をお願いします！ 🙏  viewの動線だけでなく、Controller側での対処をお願いしたいです。

YukiWatanabe824 replied:
指摘ありがとうございます。

viewとcontrollerの両方でユーザーを照合して、Viewに削除を表示するか、Controllerで削除のリクエストを受け付けるかを指定していするようにしましたが、このやり方で良いでしょうか？
Viewで`削除`を表示するor表示しないだと不正にリクエストを送られて削除される気がします。

betachelsea replied:
> Viewで削除を表示するor表示しないだと不正にリクエストを送られて削除される気がします。

おっしゃる通りかと！ Viewで導線を適切に表示して、Controllerでもしっかり不正なリクエストをまもる、というのが良いですね 👍

betachelsea commented on app/models/comment.rb:
```diff
@@ -0,0 +1,8 @@
+class Comment < ApplicationRecord
+  belongs_to :user
+  belongs_to :commentable, polymorphic: true
+  validates :user_id, presence: true
```
[nits] `user_id` の存在有無は、 `belongs_to :user` の定義の時点で必須としてくれているので、書かなくても大丈夫です。余談ですが、 `belongs_to :user` は書いてあるけど、`user_id` はあってもなくても良い、という場合は、

```ruby
belongs_to :user, optional: true
```

とします。

betachelsea commented on app/models/report.rb:
```diff
@@ -0,0 +1,10 @@
+
+class Report < ApplicationRecord
+  belongs_to :user
+  default_scope -> { order(created_at: :desc) }
```
[IMO] default_scope は、根本的な呼び出しに作用するので（意図しなくても自動的に created_at の desc の順で並ぶ）、もし逆順にしたくなった場合などに困る可能性が高いです。あまり評判はよくない、ということを共有しておきます〜 🙏 

- [Railsのdefault\_scopeは使うな、絶対（翻訳）｜TechRacho by BPS株式会社](https://techracho.bpsinc.jp/hachi8833/2021_11_04/47302)
- [\[Rails\] default\_scopeを使ったせいで泣きを見たクレイジーな困難たちを紹介するぜ！ \- Rista Tech Blog](https://blog.rista.jp/entry/2017/10/10/085120)

YukiWatanabe824 replied:
添付資料わかりやすかったです。ありがとうございます🙏

default_scopeはDBへのクエリにデフォで追加されるようになるので使い勝手が悪くなる。データの並び替えであれば取得したデータの配列を並び替えるほうがベター、という理解で良いでしょうか？

betachelsea replied:
> default_scopeはDBへのクエリにデフォで追加されるようになるので使い勝手が悪くなる。

はい、これはその解釈で良いと思います。「実装者が意図しないところで勝手に動く」=「使い勝手が悪い」 という印象です。

> データの並び替えであれば取得したデータの配列を並び替えるほうがベター

これは場合によりますが、SQLとしての `order` 自体はきちんと使いこなした方が良いと思います。orderを使うなということではなく、あくまで `default_scope` を使うことが問題(になる可能性が高い)、という認識をしていただけると良さそうです。

betachelsea commented on app/models/report.rb:
```diff
@@ -0,0 +1,10 @@
+  validates :user_id, presence: true
+  validates :content, presence: true
+  validates :title, presence: true
+  has_many :comments, as: :commentable, dependent: :destroy
```
消えた時のことを考慮された設定になっていますね！ Goodですね 👍  ✨

betachelsea commented on app/views/reports/_form.html.erb:
```diff
@@ -0,0 +1,28 @@
+  </div>
+
+  <div class="actions">
+    <%= form.hidden_field :user_id, :value => current_user.id %>
```
[MUST] 作者のuser_idをhidden_fieldで渡す仕様にすると、開発者ツールなどでここの数字をいじることで、他者になりすましての投稿が可能になってしまうので、危ないですね。誰の投稿かというのは、サーバ側(Controller)でログインユーザを確認するようにしてもらいたいです！ 🙏

betachelsea commented on app/views/reports/index.html.erb:
```diff
@@ -0,0 +1,32 @@
+  <tbody>
+    <% @reports.each do |report| %>
+      <tr>
+        <% if report.user.name == "" %>
+          <td><%= report.user.email %></td>
+        <% else %>
+          <td><%= report.user.name %></td>
+        <% end %>
```
[MUST] この `user.name` の有無を判断して出し分ける、というのが何回かプログラム中に出てきているようですね。DRYにしたいなと感じました。Userモデルにメソッドを定義してみましょう & viewの表示を差し替えてみましょう！

betachelsea commented on app/views/books/show.html.erb:
```diff
@@ -20,5 +20,42 @@
+  </thead>
+
+  <tbody>
+    <% @comments.each do |comment| %>
```
[nits] reportsとbooksで同じ内容を使っている様子？ 共通化できると良さそうですね。

参考: [レイアウトとレンダリング \- Railsガイド - 3.4 パーシャルを使う](https://railsguides.jp/layouts_and_rendering.html#%E3%83%91%E3%83%BC%E3%82%B7%E3%83%A3%E3%83%AB%E3%82%92%E4%BD%BF%E3%81%86)

betachelsea commented on app/views/comments/_new.html.erb:
```diff
@@ -0,0 +1,25 @@
+      <h2><%= pluralize(@comment.errors.count, "error") %> prohibited this comment from being saved:</h2>
+
+      <ul>
+        <% @comment.errors.each do |error| %>
```
[nits] 部分テンプレートでは、インスタンス変数 ( `@comment`、`@commentable` とか )を使用しないことをオススメします。なぜなら、その部分テンプレートを使うために必ずその名前のインスタンス変数を定義しておかねばならない、という縛りが自然とできてしまうからです。(依存性が高まる) 
ローカル変数を受け取って使えるようにしておくと便利です！

参考: [Rails　パーシャル\(部分テンプレート\)へローカル変数を渡したいとき \- Qiita](https://qiita.com/takannporo/items/a8ff93109afc3bc3bab4)

betachelsea commented on app/views/comments/_new.html.erb:
```diff
@@ -0,0 +1,25 @@
+
+  <div class="actions">
+    <%= form.hidden_field :user_id, :value => current_user.id %>
+    <%= form.hidden_field :commentable_id, :value => @commentable.id %>
+    <%= form.hidden_field :commentable_type, :value => @commentable.class %>
```
[nits] commentable_id, commentable_type は、リクエスト先のURLからも導出できそうです。

(Comment作成)
- `/reports/:report_id/comments`
- `/books/:book_id/comments`

formに露出していると、やはり勝手にユーザにいじられてURLと不整合な内容が飛んでくる可能性があるので、リクエストされたURLの内容からのみの判断にすると、迷わなくて良いですね。

betachelsea reviewed (commented):
コメントしました！ 📝

YukiWatanabe824 committed:
レビュー対応

YukiWatanabe824 committed:
レビューの対応
commentscontrollerの不要なアクションの削除 / commentのrouteを見直し /
重複処理の共通化

YukiWatanabe824 commented:
時間をかけてしまいすみません💦
修正しましたので再度確認お願いいたします。

指摘いただいた事項はすべて修正してあります。
また、指摘事項に対しての確認は各指摘に対しコメントしています。そちらもご確認お願いいたします。

betachelsea commented on app/controllers/books_controller.rb:
```diff
@@ -68,6 +72,10 @@ def set_book
   end
 
+  def set_commentable
+    @commentable = @book
```
[nits] この `@commentable` は `app/views/comments/_new.html.erb` の利用のために定義しておかないといけない、という状態になっている様子でしょうか :eyes:  ( `app/views/books/show.html.erb` を見ていて、そう考えました ) 
うまくローカル変数を使って `@commentable` の定義を無くせれば、気を配る必要のあるインスタンス変数が減って楽になるかなと思いました〜 :eyes:

YukiWatanabe824 replied:
>[nits] この @commentable は app/views/comments/_new.html.erb の利用のために定義しておかないといけない、という状態になっている様子でしょうか 👀

`app/views/comments/_new.html.erb`に使用する@commentableは不要でした💦前回ご指摘のパーシャルにローカル変数を渡す処理を使って@commentableは消去しました

betachelsea commented on db/migrate/20221025210953_create_comments.rb:
```diff
@@ -0,0 +1,12 @@
+    create_table :comments do |t|
+      t.text :content
+      t.references :user, null: false, foreign_key: true
+      t.references :commentable, polymorphic: true
```
[IMO] もしもコメントがcommentableを必ず必要とする場合は、ここにも `null: false` が必要かもしれません :eyes:  (逆にいうと単独の(commentableを持たない) comments を将来作成する予定があるなら、nullを許しておくのも案ですね。)

YukiWatanabe824 replied:
現状、単独のコメントを実装する予定はないので`null: false`を追加しました！

betachelsea commented on db/migrate/20221011213245_create_reports.rb:
```diff
@@ -0,0 +1,12 @@
+    create_table :reports do |t|
+      t.references :user, null: false, foreign_key: true
+      t.string :title
+      t.text :content
```
[nits] モデル側で `presence: true` のバリデーションをされていたので、migration時にも `null: false` をしておくとより確実かなと思いました。

YukiWatanabe824 replied:
こちら追加しました。
モデル側のバリデーションだけではSQLでの不正なクエリに対処できないのでmigrationにも`null: false`を追加する、という理解で良いでしょうか？？

betachelsea commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,41 @@
+  # Use callbacks to share common setup or constraints between actions.
+  def set_comment
+    @comment = Comment.find_by!(id: params[:id])
+  end
+
+  # Only allow a list of trusted parameters through.
+  def comment_params
+    params.require(:comment).permit(:user_id, :content, :commentable_type, :commentable_id)
```
[nits] paramsとして送られてくる情報は `:content` だけのようなので、 `:user_id`, `:commentable_type`, `:commentable_id` の許可は出さない方が良さそうです。 

許可は必要最低限に収めておくと安全です。今回はどちらにせよ `user_id, commentable_type, commentable_id` は上書きされるので問題ありませんが、そういう処理がなかったときはformに含まれる項目を開発者ツールとかでいじったときに素通りして意図しない変更がされてしまうかもしれないです。

betachelsea commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,41 @@
+    @comment = Comment.new(comment_params)
+    @comment[:user_id] = current_user.id
+    @comment[:commentable_id] = @commentable.id
+    @comment[:commentable_type] = @commentable.class
```
[nits] インスタンスに対して値を含める時は、ハッシュ的な呼び出し方よりも、次のように書くのをよく見ます〜 :eyes: 

```ruby
@comment.user_id = current_user.id
```

また、関係性がすでに定義できているので( belongs_to とか )、次のように書くとよりRailsっぽいかなと思います。

```ruby
@comment.user = current_user
@comment.commentable = @commentable
```

betachelsea commented on app/controllers/reports_controller.rb:
```diff
@@ -0,0 +1,73 @@
+
+  # PATCH/PUT /reports/1
+  def update
+    if @report.update(report_params)
```
[MUST] updateは不正なリクエストから守られてなさそうかも？と思いました。 対応お願いします！ 🙏 

以下、豆知識です 💡 
次のようにすると、そのユーザが持ってるreportからのみデータを取ってきてくれます。

```ruby
# たとえば、別のユーザが持ってるid:5の本があっても、current_userのidと紐づいてなければエラーとなる。
current_user.reports.find(params[:id])
```

`find` はデータが見つからないときは例外エラーとしてくれるので、これを使うと、自分の持っていないほかのデータを不正に呼び出そうとした時点で問題ありとしてエラーになってくれて便利です。

また、応用ですが、次のようにすると最初から user_id に current_userのidが入った状態で new されます。

```ruby
current_user.reports.new(...)
```

betachelsea commented on app/models/report.rb:
```diff
@@ -0,0 +1,9 @@
+
+class Report < ApplicationRecord
+  belongs_to :user
+  validates :user_id, presence: true
```
> [nits] user_id の存在有無は、 belongs_to :user の定義の時点で必須としてくれているので、書かなくても大丈夫です。

[nits] こちらの `user_id` も同様ですね :eyes: 
レビューをうけたときの修正のコツですが、レビュアーは同じような修正点だとひとつひとつにコメントつけられないケースも多い(コメント数が多すぎて煩雑になる問題もある)ので、ひとつ指摘があった場合、同じ指摘が無いか都度の見直しを徹底いただくとスムーズなのでオススメです！ 👍

betachelsea commented on app/models/user.rb:
```diff
@@ -28,4 +30,12 @@ def unfollow(user)
   end
+
+  def name_or_email(user)
+    if user.name == ''
+      user.email
+    else
+      user.name
+    end
```
[MUST] 引数に user がありますが、モデル内のメソッドなので、自分自身の値を参照しましょう！ 👀 

```ruby
# たとえばこんなふうなメソッドが User にあると…
def my_name 
  "私の名前は#{name}です"
end

# こんなふうに参照できます
user = User.new(name: "山田")
user.my_name #=> "私の名前は山田です"
```

また、次のイディオムを使うとかなりシンプルに書けるかもしれません 💡 参考まで！

```ruby
# presence は、空文字もしくはnilの場合、一律で nil を返します。値が入っていれば、値を返します。
x = nil
x.presence #=> nil
y = ""
y.presence #=> nil
z = "piyo"
z.presence #=> "piyo"

# || で繋げると、左側から表記し、nilの場合は無視されます。
a = nil
b = "hoge"
a || b #=> "hoge"


```

betachelsea commented on app/views/comments/show.html.erb:
```diff
@@ -0,0 +1,27 @@
+
+<p>
+  <strong><%= Comment.human_attribute_name(:user) %>:</strong>
+  <% if @comment.user.name == "" %>
```
[MUST] せっかくなのでここも `name_or_email` メソッドを使いましょう！ 👍  他にも適用できそうなところを見つけたので、全体を見直してみてください。
PullRequestの 「Files changed」タブを活用して、都度見直す、というのをぜひ習慣づけていってほしいです！

betachelsea reviewed (commented):
修正ありがとうございます！コメント追加しました〜 📝

YukiWatanabe824 committed:
コメント生成フォーム作成用の@commentable削除

@commentableは@report/@bookを代入しており、コメント生成フォームパーシャル呼び出し時にローカル変数として代入すれば済む。
不要なインスタンス変数を除去し、依存性を下げるため。

YukiWatanabe824 committed:
reports/commentsのマイグレーションを修正

reportのtilte,contentにnull制約追加、commentsのcommentableにnull制約を追加
モデル側でpresence:trueを設定していたが、DB側の制約がなく安全性向上のため追加

YukiWatanabe824 committed:
comment paramsの項目を変更

content以外はparamsから取得しないため削除

YukiWatanabe824 committed:
データの代入方法を修正

Railsらしい記述に修正

YukiWatanabe824 committed:
updateの不正リクエスト対策を追加

ログインユーザーによるupdateか判定する処理を追加

YukiWatanabe824 committed:
report Create時のインスタンス生成を修正

current_userに紐づけてインスタンスを生成し、DRYに

YukiWatanabe824 committed:
userに関するvaridationを削除

belongs_toの段階でuser必須となるため

YukiWatanabe824 committed:
name_or_emailを修正

インスタンスを参照するように修正

YukiWatanabe824 committed:
ユーザ名/メアドを呼び出す処理を差替え

YukiWatanabe824 commented:
質問への回答ありがとうございました！

修正し再度プルリクを送ります。

前回コミットがざっくりしていて確認しづらかったかと思いまして、
修正箇所ごとにコミットを分けてみました。

ご確認お願いいたします。

betachelsea commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,40 @@
+  # POST /comments
+  def create
+    @comment = Comment.new(comment_params)
+    @comment.commentable = @commentable
```
[IMO] `@commentable` とインスタンス変数になっていますが、インスタンス変数は結構スコープが広いので、どこに定義があるのか探るのがちょっと難しいです。継承先の `set_commentable` メソッドで代入しているようですが、この CommentsController 単体で見ると状況が把握できないため、この例でも難しいなと感じました。

案として、メソッドとして定義しておき、継承時にオーバーライドしないとエラーがでるようにしておくというやり方もあるよ〜というのをご紹介しておきます 🙏 

```ruby
# 諸々一部省略
class CommentsController < ApplicationController

  def create
    @comment = Comment.new(comment_params)
    @comment.commentable = commentable # メソッド呼び出し
    # ...略...
  end

  private
  
  # 参考: [class NotImplementedError \(Ruby 3\.1 リファレンスマニュアル\)](https://docs.ruby-lang.org/ja/latest/class/NotImplementedError.html)
  def commentable
    raise NotImplementedError
  end
  
end
```

```ruby
class Reports::CommentsController < CommentsController

  private

  # 継承時にオーバーライドしておかないと例外が出るのでちゃんと書く
  def commentable
    # ※遅延初期化のイディオム 複数回呼び出したときに都度 find しなくて良くなる
    @commentable ||= Report.find(params[:report_id])
  end
end
```

betachelsea commented on app/controllers/reports_controller.rb:
```diff
@@ -0,0 +1,72 @@
+
+  # Only allow a list of trusted parameters through.
+  def report_params
+    params.require(:report).permit(:user_id, :title, :content)
```
[nits] こちらの `user_id` も、Booksの時の考え方と同様に不要そうです 👀

betachelsea commented on app/controllers/reports_controller.rb:
```diff
@@ -0,0 +1,72 @@
+# frozen_string_literal: true
+
+class ReportsController < ApplicationController
+  before_action :set_report, only: %i[show edit destroy]
+  before_action :set_updated_report, only: %i[update]
```
[IMO] こちら、 edit ,destroy も「自分自身の日報」のみを対象とした操作なので、 `set_updated_report` を適用するのも一案かもしれません。そうなると、edit, destroy での `@report.user != current_user` の確認は要らなくて済みそうです。 ただしその場合は、丁寧なエラー返答はなくなります。しかしながら、「他人の日報」にイタズラしようとした人に丁寧な案内が必要なのか？という意見があるのも確かですね〜 👀

betachelsea commented on app/views/books/show.html.erb:
```diff
@@ -20,5 +20,12 @@
 
+<%= render partial: 'comments/comments', locals: { comments: @comments, commentable: @book} %>
+
+<br>
```
[nits] `<br>` は文章を改行するための要素なので、こういう場所に唐突に出てくるのはちょっと気になりました〜 👀  レイアウトはcssを使っていきましょう！ 💪

betachelsea reviewed (approved):
修正ありがとうございました！ 確認OKです 👍 
少しコメントしました〜 📝

YukiWatanabe824 committed:
reportのURLパラメータ修正、使用メソッド修正

YukiWatanabe824 committed:
HTMLタグの修正

HTMLタグはデザインのために使わない、の方針に則り、brタグを削除
